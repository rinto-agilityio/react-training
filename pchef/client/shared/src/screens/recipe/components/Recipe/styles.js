// Libs
import { StyleSheet } from 'react-native'

// Themes
import { COLORS, METRICS, FONTS } from '../../../../themes'

const styles = StyleSheet.create({
  wrapper: {
    display: 'flex',
    padding: METRICS.padding.sm,
  },

  // Title
  title: {
    color: COLORS.black,
    fontWeight: 'bold',
    marginBottom: METRICS.margin.lg,
  },
  smallTitle: {
    fontSize: FONTS.fontSize.medium,
  },
  mediumTitle: {
    fontSize: FONTS.fontSize.large,
  },
  largeTitle: {
    fontSize: FONTS.fontSize.extraLarge,
  },

  // Description
  smallDescription: {
    fontSize: FONTS.fontSize.small,
  },
  mediumDescription: {
    fontSize: FONTS.fontSize.medium,
  },
  largeDescription: {
    fontSize: FONTS.fontSize.large,
  },

  // Ingredients
  descriptionIngredients: {
    color: COLORS.black,
    marginBottom: METRICS.margin.md,
  },
  ingredients: {
    borderBottomColor: COLORS.baseGray,
    borderBottomWidth: 1,
  },

  // Directions
  directions: {
    marginTop: METRICS.margin.lg,
  },
  wrapperDescriptions: {
    lineHeight: METRICS.lineHeight.sm,
    width: '85%',
  },
  mediumWrapperDescriptions: {
    width: '75%',
  },
  descriptionDirections: {
    color: COLORS.white,
    backgroundColor: COLORS.black,
    fontWeight: 'bold',
    paddingLeft: METRICS.padding.lg,
    paddingRight: METRICS.padding.sm,
  },

  // Button
  button: {
    backgroundColor: COLORS.black,
    borderColor: COLORS.black,
    minWidth: METRICS.largeIcon,
  },
  smallButton: {
    borderRadius: METRICS.borderRadius.xs,
    padding: 0,
  },
  mediumButton: {
    borderRadius: METRICS.borderRadius.md,
    padding: METRICS.padding.sm,
  },
  largeButton: {
    borderRadius: METRICS.borderRadius.md,
    padding: METRICS.padding.sm,
  },
  titleBtn: {
    fontWeight: 'bold',
  },
  smallTitleBtn: {
    fontSize: FONTS.fontSize.medium,
  },
  mediumTitleBtn: {
    fontSize: FONTS.fontSize.large,
  },
  largeTitleBtn: {
    fontSize: FONTS.fontSize.extraLarge,
  },

  // Style for text in tag component
  customTextStyle: {
    marginBottom: '5px',
  },
})

export default styles
