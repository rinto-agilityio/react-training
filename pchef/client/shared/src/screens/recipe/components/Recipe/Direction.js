// @flow
// add flow above to fix for using flow with React.memo

// Libs
import React, { memo } from 'react'
import { Text } from 'react-native'

// Styles
import styles from './styles'

// Themes
import { METRICS } from '../../../../themes'

// Components
import Button from '../../../../components/Button'
import Wrapper from '../../../../layout/Wrapper'

type Props = {
  item: {
    step: number,
    title: string,
    id: string,
  },
  size: string,
  customDescription?: {},
  onSelectStep?: (id: string) => void,
}

const Direction = ({
  item = {},
  size = 'medium',
  customDescription,
  onSelectStep = () => {},
}: Props) => (
  <Wrapper
    direction="row"
    childPosition="middle"
    customStyles={{
      marginBottom: METRICS.largeMargin,
    }}
  >
    <Button
      title={item.step ? item.step.toString() : ''}
      buttonStyle={[styles.button, styles[`${size}Button`]]}
      titleStyle={[styles.titleBtn, styles[`${size}TitleBtn`]]}
      onPress={() => onSelectStep(item.id)}
    />
    <Text style={[styles.wrapperDescriptions, styles[`${size}WrapperDescriptions`]]}>
      <Text
        style={[
          styles.descriptionDirections,
          styles[`${size}Description`],
          customDescription,
        ]}
        onPress={onSelectStep}
      >
        {item.title}
      </Text>
    </Text>
  </Wrapper>
)

Direction.defaultProps = {
  customDescription: {},
  onSelectStep: () => {},
}

export default memo<Props>(Direction)
