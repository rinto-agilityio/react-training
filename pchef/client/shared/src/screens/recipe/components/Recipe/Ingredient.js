// @flow
// add flow above to fix for using flow with React.memo

// Libs
import React, { memo } from 'react'
import { Text } from 'react-native'

// Styles
import styles from './styles'

type Props = {
  item: string,
  size: string,
  customDescription?: {},
}

const Ingredient = ({
  item = '',
  size = 'medium',
  customDescription,
}: Props) => (
  <Text
    style={[
      styles.descriptionIngredients,
      styles[`${size}Description`],
      customDescription,
    ]}
  >
    {item ? item.trim() : ''}
  </Text>
)

Ingredient.defaultProps = {
  customDescription: {},
}

export default memo<Props>(Ingredient)
