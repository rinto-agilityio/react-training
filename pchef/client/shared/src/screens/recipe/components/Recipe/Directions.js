// Libs
import React from 'react'
import { Text, View } from 'react-native'

// Styles
import styles from './styles'

// Components
import Direction from './Direction'

type Props = {
  steps: Array<{
    step: number,
    title: string,
    id: string,
  }>,
  size: string,
  customDirections?: {},
  customTitle?: {},
  onSelectStep?: () => void,
}

const Directions = ({
  steps = [],
  size = 'medium',
  customDirections,
  customTitle,
  onSelectStep,
}: Props) => (
  <View
    style={[
      styles.directions,
      styles[`${size}Directions`],
      customDirections,
    ]}
  >
    <Text
      style={[
        styles.title,
        styles[`${size}Title`],
        customTitle,
      ]}
    >
      How to cook
    </Text>

    {steps.map(item => (
      <Direction
        item={item}
        size={size}
        onSelectStep={onSelectStep}
        key={item.step}
      />
    ))}
  </View>
)

Directions.defaultProps = {
  customDirections: {},
  customTitle: {},
  onSelectStep: () => {},
}

export default Directions
