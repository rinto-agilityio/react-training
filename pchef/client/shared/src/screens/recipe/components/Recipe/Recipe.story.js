// Libs
import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

// Components
import Recipe from '.'

// Mocks
import { recipes } from '../../../../mocks'

storiesOf('Recipe', module)
  .add('small', () => (
    <Recipe
      recipe={recipes[0]}
      size="small"
      onSelectStep={action('clicked')}
    />
  ))
  .add('medium', () => (
    <Recipe
      recipe={recipes[0]}
      size="medium"
      onSelectStep={action('clicked')}
    />
  ))
  .add('large', () => (
    <Recipe
      recipe={recipes[0]}
      size="large"
      onSelectStep={action('clicked')}
    />
  ))
