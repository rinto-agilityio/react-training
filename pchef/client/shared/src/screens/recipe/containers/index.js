import { compose } from 'react-apollo'

// GraphQL
import { recipeDetail } from '../../../graphql/recipe'
import { getAllRecipeSteps } from '../../../graphql/recipe-step'
import Recipe from '../components/Recipe'

export default compose(getAllRecipeSteps, recipeDetail)(Recipe)
