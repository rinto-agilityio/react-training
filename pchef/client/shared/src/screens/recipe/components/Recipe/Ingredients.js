// @flow
// add flow above to fix for using flow with React.memo

// Libs
import React, { memo } from 'react'
import { Text, View } from 'react-native'

// Styles
import styles from './styles'

// Helpers
import { formatStringToArray } from '../../../../helpers/utils'

// Constants
import { SEPARATOR_SPLIT_STRING } from '../../../../constants'

// Components
import Ingredient from './Ingredient'
import Wrapper from '../../../../layout/Wrapper'

type Props = {
  description: string,
  size: string,
  customIngredients?: {},
  customTitle?: {},
  onClose?: (item: string) => void,
  disabled?: boolean,
  customWrapperTagStyles: Object,
}

const Ingredients = ({
  description = '',
  size = 'medium',
  customIngredients,
  customTitle,
  onClose = () => {},
  disabled = false,
  customWrapperTagStyles,
}: Props) => {
  const ingredients = (description && formatStringToArray(description, SEPARATOR_SPLIT_STRING)) || []
  const ingredientAmount = ingredients.length

  return (
    <View
      style={[
        styles.ingredients,
        styles[`${size}Ingredients`],
        customIngredients,
      ]}
    >
      <Text
        style={[
          styles.title,
          styles[`${size}Title`],
          customTitle,
        ]}
      >
        What you will need
      </Text>
      <Wrapper customStyles={customWrapperTagStyles}>
        {
          ingredientAmount > 0 ?
            ingredients.map((item, index) => (
              <Ingredient
                item={item}
                key={`${item}_${index}`}
                onClose={item => onClose(item)}
                disabled={disabled}
              />
            )) : <Text>No ingredients to show</Text>
        }
      </Wrapper>
    </View>
  )
}

Ingredients.defaultProps = {
  customIngredients: {},
  customTitle: {},
}

export default memo<Props>(Ingredients)
