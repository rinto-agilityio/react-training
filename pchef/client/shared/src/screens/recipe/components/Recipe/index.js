// Libs
import React, { useState } from 'react'
import { View } from 'react-native'

// Styles
import styles from './styles'

// Components
import Ingredients from './Ingredients'
import Directions from './Directions'
import Loading from '../../../../components/Loading'
import Modal from '../../../../components/Modal'
import Error from '../../components/Error'

// Helper
import { customError } from '../../../../helpers/utils'

type Props = {
  getRecipe: {
    description: string,
    id: string,
  },
  size: string,
  onSelectStep: (id: string) => void,
  loading: boolean,
  error: Object,
  recipeSteps: Array<{
    step: number,
    title: string
  }>,
  history: Object,
}

const Recipe = ({
  getRecipe = {
    description: '',
    id: '',
  },
  size = 'medium',
  onSelectStep,
  loading,
  error,
  recipeSteps = [{
    step: 1,
    title: '',
  }],
  history,
}: Props) => {
  const [visible, setVisible] = useState(true)

  if (loading) {
    return <Loading />
  }

  const handleNavigateLogin = () => {
    setVisible(false)
    history.push('/login')
  }

  if (error) {
    return (
      <Modal
        visible={visible}
        onDismiss={() => handleNavigateLogin()}
        onSubmit={() => handleNavigateLogin()}
        size={size}
      >
        <Error message={customError(error.graphQLErrors)} size="medium" />
      </Modal>
    )
  }

  const {
    description,
    id,
  } = getRecipe

  return (
    <View style={[styles.wrapper, styles[`${size}Wrapper`]]}>
      <Ingredients
        description={description}
        size={size}
      />
      <Directions
        steps={recipeSteps}
        size={size}
        onSelectStep={() => onSelectStep(id)}
      />
    </View>
  )
}

export default Recipe
