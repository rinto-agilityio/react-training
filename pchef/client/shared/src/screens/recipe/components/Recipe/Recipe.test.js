// Components
import Recipe from '.'
import Directions from './Directions'
import Direction from './Direction'
import Ingredients from './Ingredients'
import Ingredient from './Ingredient'

// Mocks
import { recipe } from '../../../../mocks/recipe'

describe('Recipe by step', () => {
  const recipeProps = {
    recipe,
    onSelectStep: jest.fn(),
  }

  const recipeComponent = mount(<Recipe {...recipeProps} />)
  const directionsComponent = mount(<Directions onSelectStep={jest.fn()} />)
  const directionComponent = shallow(<Direction onSelectStep={jest.fn()} />)
  const ingredientsComponent = mount(<Ingredients />)
  const ingredientComponent = shallow(<Ingredient />)

  it('Renders correctly recipe commponent', () => {
    expect(recipeComponent).toMatchSnapshot()
  })

  it('Render recipe component with size medium', () => {
    recipeComponent.setProps({
      size: 'medium',
    })
    expect(recipeComponent).toMatchSnapshot()
  })

  it('Function props of recipe component should be defined', () => {
    Recipe.defaultProps.onSelectStep()
    expect(Recipe.defaultProps.onSelectStep).toBeDefined()
  })

  it('Renders correctly directions commponent', () => {
    expect(directionsComponent).toMatchSnapshot()
  })

  it('Render directions component with size medium', () => {
    directionsComponent.setProps({
      size: 'medium',
    })
    expect(directionsComponent).toMatchSnapshot()
  })

  it('Render directions component with steps array', () => {
    directionsComponent.setProps({
      steps: recipe.steps,
    })
    expect(directionsComponent).toMatchSnapshot()
  })

  it('Function props of directions component should be defined', () => {
    Directions.defaultProps.onSelectStep()
    expect(Directions.defaultProps.onSelectStep).toBeDefined()
  })

  it('Renders correctly direction commponent', () => {
    expect(directionComponent).toMatchSnapshot()
  })

  it('Render direction component with size medium', () => {
    directionsComponent.setProps({
      size: 'medium',
    })
    expect(directionComponent).toMatchSnapshot()
  })

  it('Render direction component with step info', () => {
    directionsComponent.setProps({
      item: recipe.steps[0],
    })
    expect(directionComponent).toMatchSnapshot()
  })

  it('Function props of direction component should be defined', () => {
    Direction.defaultProps.onSelectStep()
    expect(Direction.defaultProps.onSelectStep).toBeDefined()
  })

  it('Renders correctly ingredients commponent', () => {
    expect(ingredientsComponent).toMatchSnapshot()
  })

  it('Render ingredients component with size medium', () => {
    ingredientsComponent.setProps({
      size: 'medium',
    })
    expect(ingredientsComponent).toMatchSnapshot()
  })

  it('Render directions component with recipe description', () => {
    ingredientsComponent.setProps({
      description: recipe.description,
    })
    expect(ingredientsComponent).toMatchSnapshot()
  })

  it('Renders correctly ingredient component', () => {
    expect(ingredientComponent).toMatchSnapshot()
  })

  it('Render ingredient component with size medium', () => {
    ingredientComponent.setProps({
      size: 'medium',
    })
    expect(ingredientComponent).toMatchSnapshot()
  })

  it('Render ingredient component with text', () => {
    ingredientComponent.setProps({
      item: 'recipe description',
    })
    expect(ingredientComponent).toMatchSnapshot()
  })
})
