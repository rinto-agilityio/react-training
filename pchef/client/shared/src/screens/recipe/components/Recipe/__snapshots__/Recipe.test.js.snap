// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe by step Render Recipe component with loading true 1`] = `
<div
  className="css-view-1dbjc4n r-justifyContent-1777fci"
  style={
    Object {
      "WebkitBoxFlex": 1,
      "WebkitFlexBasis": "0%",
      "WebkitFlexGrow": 1,
      "WebkitFlexShrink": 1,
      "flexBasis": "0%",
      "flexGrow": 1,
      "flexShrink": 1,
      "msFlexNegative": 1,
      "msFlexPositive": 1,
      "msFlexPreferredSize": "0%",
    }
  }
>
  <div
    aria-valuemax="1"
    aria-valuemin="0"
    className="css-view-1dbjc4n r-alignItems-1awozwy r-justifyContent-1777fci"
    role="progressbar"
  >
    <div
      className="css-view-1dbjc4n r-animationDuration-17bb2tj r-animationIterationCount-1muvv40 r-animationKeyframes-127358a r-animationTimingFunction-1ldzwu0 r-height-z80fyv r-width-19wmn03"
    >
      <svg
        height="100%"
        viewBox="0 0 32 32"
        width="100%"
      >
        <circle
          cx="16"
          cy="16"
          fill="none"
          r="14"
          strokeWidth="4"
          style={
            Object {
              "opacity": 0.2,
              "stroke": "#c5c5c5",
            }
          }
        />
        <circle
          cx="16"
          cy="16"
          fill="none"
          r="14"
          strokeWidth="4"
          style={
            Object {
              "stroke": "#c5c5c5",
              "strokeDasharray": 80,
              "strokeDashoffset": 60,
            }
          }
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Recipe by step Render direction component with size medium 1`] = `ShallowWrapper {}`;

exports[`Recipe by step Render direction component with step info 1`] = `ShallowWrapper {}`;

exports[`Recipe by step Render directions component with recipe description 1`] = `ReactWrapper {}`;

exports[`Recipe by step Render directions component with size medium 1`] = `ReactWrapper {}`;

exports[`Recipe by step Render directions component with steps array 1`] = `ReactWrapper {}`;

exports[`Recipe by step Render ingredient component with size medium 1`] = `ShallowWrapper {}`;

exports[`Recipe by step Render ingredient component with text 1`] = `ShallowWrapper {}`;

exports[`Recipe by step Render ingredients component with size medium 1`] = `ReactWrapper {}`;

exports[`Recipe by step Render recipe component with size medium 1`] = `ReactWrapper {}`;

exports[`Recipe by step Renders correctly direction commponent 1`] = `ShallowWrapper {}`;

exports[`Recipe by step Renders correctly directions commponent 1`] = `ReactWrapper {}`;

exports[`Recipe by step Renders correctly ingredient component 1`] = `ShallowWrapper {}`;

exports[`Recipe by step Renders correctly ingredients commponent 1`] = `ReactWrapper {}`;

exports[`Recipe by step Renders correctly recipe commponent 1`] = `ReactWrapper {}`;
